### ISSUE TITLE
How to generate the answer JSON file for test?

### ISSUE THREAD

**Issue by BA3000** [hexiang-hu/answer_embedding#2]
The VQA challenge requires uploading answer JSON file to their server for testing. However, I did not find related functions in your code, except the functions in `dump_vqa_eval_json.py`. Maybe I missed them ?

**Comment by BA3000** [comment]
Me again. I have found the related functions for generating answer JSON file, and successfully generated one for `test-dev` split. I uploaded the file to the VQA server for evaluation and got the results listed in the article. Here is how I did that.
  The related functions are `test`, which is defined in `train_vqa_embedding.py`, and `data.get_loader`, which is defined in `ansemb/dataset/vqa.py`. For convenience, we shall copy the `train_vqa_embedding.py`, rename the copy as `vqa_test-dev.py` and modify it. First, we need to delete all the accuracy calculating code in `test` function. `test` function was originally used for calculating accuracy on `val` split. The `test` function works in this way. First, it feeds the batched question and image feature into the trained model, gets answers from the model, then calculates accuracy comparing them to ground truth answer. The answers of `test` split are not provided, or else, we don't have to upload JSON file to their server for evaluation, so we do not need the accuracy calculation part. After doing so, the resulted code is supposed to be like this:

```python
def test(context_net, answer_net, loader, tracker, args, prefix='', epoch=0):
    context_net.eval()
    answer_net.eval()
    ans_ids, que_ids = [], []
    masks = []

    tq = tqdm(loader, desc='{} E{:03d}'.format(prefix, epoch), ncols=0)

    var_params = {'volatile': True, 'requires_grad': False}
    if args.answer_embedding == 'RNN':
        answer_var, answer_len = loader.dataset._get_answer_sequences(
            range(cfg.TEST.max_answer_index))
    else:
        answer_var, answer_len = loader.dataset._get_answer_vectors(
            range(cfg.TEST.max_answer_index))

    answer_var = Variable(answer_var.cuda(), **var_params)
    answer_embedding = answer_net.forward(answer_var, answer_len)

    for v, q, idx, q_len in tq:
        v = Variable(v.cuda(), **var_params)
        q = Variable(q.cuda(), **var_params)
        q_len = Variable(q_len.cuda(), **var_params)

        context_embedding = context_net(v, q, q_len)

        predicts = cosine_sim(
            context_embedding, answer_embedding) / args.loss_temperature  # temperature

        # collect stats
        _, _ans_ids = predicts.data.cpu().max(dim=1)
        ans_ids.append(_ans_ids.view(-1))
        que_ids.append(idx.view(-1).clone())

    return masks, ans_ids, que_ids
```

  I forgot to remove the `masks` part, all we need to return is `ans_ids` and `que_ids`. Also, remember to modify the `for v, q, idx, q_len in tq:` part, answers are not provided in `test` split.
  Then, modify the `main` part, change `val_loader = data.get_loader(word2vec, val=True)` to `test_loader = data.get_loader(word2vec, test=True)`, and replace all `val_loader` with `test_loader`. After doing all of these, we are almost done. Run the `vqa_test-dev.py` and we shall get the output file which is named like `***.pth` in `outputs` folder. We need to use `dump_vqa_eval_json.py` for generating the answer JSON file we want. The command is like this: `python dump_vqa_eval_json.py --input_path ../outputs/SAN_BoW_vqa_batch_softmax_embedding_2019-05-06_12\:47\:20.pth --output_path ../outputs/test.json --input_json ../data/vqa2/v2_OpenEnded_mscoco_test-dev2015_questions.json`. We shall be able to find `test.json`, the answer JSON file, in `outputs` folder when the command is finished.

The overall code of `vqa_test.py` shall be looks like:

```python
from ansemb.vector import Vector
from ansemb.utils import cosine_sim
import ansemb.utils as utils
import ansemb.models.embedding as model
import ansemb.dataset.vqa as data
from ansemb.config import cfg, set_random_seed, update_train_configs
import sys
import os.path
import math
import json
import numpy as np
import random

import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torch.autograd import Variable
import torch.backends.cudnn as cudnn
from tqdm import tqdm
import copy
import argparse
import time

cudnn.enabled = True
cudnn.benchmark = True


parser = argparse.ArgumentParser()
parser.add_argument('--gpu_id', default=0, type=int)
parser.add_argument('--finetune', action='store_true')
parser.add_argument('--batch_size', default=128, type=float)
parser.add_argument('--max_negative_answer', default=12000, type=int)
parser.add_argument('--answer_batch_size', default=3000, type=int)
parser.add_argument('--loss_temperature', default=0.01, type=float)
parser.add_argument('--pretrained_model', default=None, type=str)
parser.add_argument('--context_embedding', default='SAN',
                    choices=['SAN', 'BoW'])
parser.add_argument('--answer_embedding', default='BoW',
                    choices=['BoW', 'RNN'])
parser.add_argument('--name', default=None, type=str)
args = parser.parse_args()

# fix random seed
set_random_seed(cfg.seed)


def test(context_net, answer_net, loader, tracker, args, prefix='', epoch=0):
    context_net.eval()
    answer_net.eval()
    ans_ids, que_ids = [], []
    masks = []

    tq = tqdm(loader, desc='{} E{:03d}'.format(prefix, epoch), ncols=0)

    var_params = {'volatile': True, 'requires_grad': False}
    if args.answer_embedding == 'RNN':
        answer_var, answer_len = loader.dataset._get_answer_sequences(
            range(cfg.TEST.max_answer_index))
    else:
        answer_var, answer_len = loader.dataset._get_answer_vectors(
            range(cfg.TEST.max_answer_index))

    answer_var = Variable(answer_var.cuda(), **var_params)
    answer_embedding = answer_net.forward(answer_var, answer_len)

    for v, q, idx, q_len in tq:
        v = Variable(v.cuda(), **var_params)
        q = Variable(q.cuda(), **var_params)
        q_len = Variable(q_len.cuda(), **var_params)

        context_embedding = context_net(v, q, q_len)

        predicts = cosine_sim(
            context_embedding, answer_embedding) / args.loss_temperature  # temperature

        # collect stats
        _, _ans_ids = predicts.data.cpu().max(dim=1)
        ans_ids.append(_ans_ids.view(-1))
        que_ids.append(idx.view(-1).clone())

    return masks, ans_ids, que_ids


def main(args):
    if args.name is None:
        from datetime import datetime
        name = args.context_embedding+"_"+args.answer_embedding + \
            "_vqa_batch_softmax_embedding_"+datetime.now().strftime("%Y-%m-%d_%H:%M:%S")
        name = (name + '_finetune') if args.finetune else name
    else:
        name = args.context_embedding+"_"+args.answer_embedding + \
            "_vqa_batch_softmax_embedding_"+args.name

    output_filepath = os.path.join(cfg.output_path, '{}.pth'.format(name))
    print('Output data would be saved to {}'.format(output_filepath))

    word2vec = Vector()
    train_loader = data.get_loader(word2vec, train=True)
    test_loader = data.get_loader(word2vec, test=True)

    question_word2vec = word2vec._prepare(train_loader.dataset.token_to_index)

    if args.context_embedding == 'SAN':
        context_net = model.StackedAttentionEmbedding(
            train_loader.dataset.num_tokens,
            question_word2vec).cuda()
    elif args.context_embedding == 'BoW':
        context_net = model.VisualSemanticEmbedding(
            train_loader.dataset.num_tokens,
            question_word2vec).cuda()
    else:
        raise TypeError('Unsupported Context Model')

    if args.answer_embedding == 'BoW':
        answer_net = model.MLPEmbedding(train_loader.dataset.vector.dim).cuda()
    elif args.answer_embedding == 'RNN':
        answer_net = model.RNNEmbedding(train_loader.dataset.vector.dim).cuda()
    else:
        raise TypeError('Unsupported Answer Model')

    print('Context Model:')
    print(context_net)

    print('Answer Model:')
    print(answer_net)

    if args.pretrained_model is not None:
        states = torch.load(args.pretrained_model)
        answer_state, context_state = states['answer_net'], states['context_net']

        answer_net.load_state_dict(answer_state)
        context_net.load_state_dict(context_state)

    tracker = utils.Tracker()
    if args.pretrained_model is not None:
        _, ans_ids, que_ids = test(
            context_net, answer_net, test_loader, tracker, args, prefix='test', epoch=-1)

        results = {'name': name,
                   'eval': [{'answer_ids': ans_ids, 'question_ids': que_ids}],
                   'vocab': {'answer_to_index': train_loader.dataset.answer_to_index,
                             'index_to_answer': train_loader.dataset.index_to_answer}}
        print('* Dumpping output to: {}'.format(output_filepath))
        torch.save(results, output_filepath)

        if not args.finetune:
            raise ValueError('Testing Finished')

if __name__ == '__main__':
    torch.cuda.set_device(args.gpu_id)
    print(args.__dict__)
    print(cfg)
    update_train_configs(args)
    main(args)
```
  Generating answer JSON file for VizWiz or VG challenges are quite similar, so I shall stop here, and let y'all figure it out by yourselves.
